File: app/orchestration/templates.py
"""
Guidance templates — versioned, deterministic mappings of intents to
ordered step blueprints.

Phase 3 ships an in-code default registry. Phase 7 layers JSON loading
on top so non-developers can author new flows without touching Python.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path

from app.state.session import ActionVerb


@dataclass(frozen=True)
class StepBlueprint:
    """Recipe for one step of guidance."""

    action: ActionVerb
    target_query: str
    instruction_key: str = "guidance.click_target"
    fallback_window: str | None = None


@dataclass(frozen=True)
class GuidanceTemplate:
    """Recipe for guiding the user through one intent end-to-end."""

    intent: str
    confirmation_action_key: str
    steps: tuple[StepBlueprint, ...]
    expected_window: str | None = None
    aliases: tuple[str, ...] = field(default_factory=tuple)


_DEFAULT_TEMPLATES: tuple[GuidanceTemplate, ...] = (
    GuidanceTemplate(
        intent="open_folder",
        confirmation_action_key="confirm.open_folder",
        expected_window="Explorer",
        steps=(
            StepBlueprint(action="double_click", target_query="{target}", instruction_key="guidance.double_click_target", fallback_window="Explorer"),
        ),
    ),
    GuidanceTemplate(
        intent="move_file",
        confirmation_action_key="confirm.move_file",
        expected_window="Explorer",
        steps=(
            StepBlueprint(action="locate", target_query="{target}", instruction_key="guidance.locate_target"),
            StepBlueprint(action="right_click", target_query="{target}", instruction_key="guidance.right_click_target"),
            StepBlueprint(action="click", target_query="Cortar", instruction_key="guidance.click_target"),
        ),
    ),
    GuidanceTemplate(
        intent="delete_file",
        confirmation_action_key="confirm.delete_file",
        expected_window="Explorer",
        steps=(
            StepBlueprint(action="click", target_query="{target}", instruction_key="guidance.click_target"),
            StepBlueprint(action="locate", target_query="Suprimir", instruction_key="guidance.locate_target"),
        ),
    ),
    GuidanceTemplate(
        intent="open_browser",
        confirmation_action_key="confirm.open_browser",
        expected_window="Chrome",
        steps=(
            StepBlueprint(action="locate", target_query="Chrome", instruction_key="guidance.locate_target"),
        ),
    ),
    GuidanceTemplate(
        intent="search_web",
        confirmation_action_key="confirm.search_web",
        expected_window="Chrome",
        steps=(
            StepBlueprint(action="click", target_query="Nueva pestaña", instruction_key="guidance.click_target"),
            StepBlueprint(action="type", target_query="Buscar", instruction_key="guidance.type_in_target"),
        ),
    ),
    GuidanceTemplate(
        intent="open_url",
        confirmation_action_key="confirm.open_url",
        expected_window="Chrome",
        steps=(
            StepBlueprint(action="click", target_query="Nueva pestaña", instruction_key="guidance.click_target"),
            StepBlueprint(action="type", target_query="Buscar", instruction_key="guidance.type_in_target"),
        ),
    ),
    GuidanceTemplate(
        intent="compose_email",
        confirmation_action_key="confirm.compose_email",
        expected_window="Gmail",
        steps=(
            StepBlueprint(action="click", target_query="Redactar", instruction_key="guidance.click_target"),
        ),
    ),
    GuidanceTemplate(
        intent="read_inbox",
        confirmation_action_key="confirm.read_inbox",
        expected_window="Gmail",
        steps=(
            StepBlueprint(action="click", target_query="Bandeja de entrada", instruction_key="guidance.click_target"),
        ),
    ),
    GuidanceTemplate(
        intent="reply_message",
        confirmation_action_key="confirm.reply_message",
        expected_window="Gmail",
        steps=(
            StepBlueprint(action="click", target_query="Responder", instruction_key="guidance.click_target"),
        ),
    ),
    GuidanceTemplate(
        intent="open_word_document",
        confirmation_action_key="confirm.open_word_document",
        expected_window="Word",
        steps=(
            StepBlueprint(action="locate", target_query="Word", instruction_key="guidance.locate_target"),
        ),
    ),
    GuidanceTemplate(
        intent="print_document",
        confirmation_action_key="confirm.print_document",
        expected_window="Word",
        steps=(
            StepBlueprint(action="click", target_query="Imprimir", instruction_key="guidance.click_target"),
        ),
    ),
)


class TemplateRegistry:
    """Lookup of intent name → GuidanceTemplate."""

    def __init__(self, templates: tuple[GuidanceTemplate, ...] = ()) -> None:
        self._by_intent: dict[str, GuidanceTemplate] = {}
        for tpl in templates:
            self.register(tpl)

    def register(self, template: GuidanceTemplate) -> None:
        self._by_intent[template.intent] = template

    def get(self, intent: str) -> GuidanceTemplate | None:
        return self._by_intent.get(intent)

    def known_intents(self) -> tuple[str, ...]:
        return tuple(sorted(self._by_intent.keys()))

    @classmethod
    def from_json_dir(cls, root: Path) -> "TemplateRegistry":
        """Load every `*.json` template under `root` and merge with defaults."""
        registry = default_registry()
        if not root.exists():
            return registry
        for path in sorted(root.glob("*.json")):
            try:
                payload = json.loads(path.read_text(encoding="utf-8"))
            except json.JSONDecodeError:
                continue
            for intent_name, body in payload.items():
                steps = tuple(
                    StepBlueprint(
                        action=step["action"],
                        target_query=step.get("target_query", "{target}"),
                        instruction_key=step.get("instruction_key", "guidance.click_target"),
                        fallback_window=step.get("fallback_window"),
                    )
                    for step in body.get("steps", [])
                )
                registry.register(
                    GuidanceTemplate(
                        intent=intent_name,
                        confirmation_action_key=body.get("confirmation_action_key", f"confirm.{intent_name}"),
                        expected_window=body.get("expected_window"),
                        steps=steps,
                        aliases=tuple(body.get("aliases", ())),
                    )
                )
        return registry


def default_registry() -> TemplateRegistry:
    """Return a fresh registry pre-populated with the in-code defaults."""
    return TemplateRegistry(_DEFAULT_TEMPLATES)

File: src-tauri/src/orchestration/templates.rs
use std::collections::HashMap;
use std::path::Path;

use serde::Deserialize;

use crate::orchestration::state::ActionVerb;

#[derive(Debug, Clone)]
pub struct StepBlueprint {
    pub action: ActionVerb,
    pub target_query: String,
    pub instruction_key: String,
    pub fallback_window: Option<String>,
}

#[derive(Debug, Clone)]
pub struct GuidanceTemplate {
    pub intent: String,
    pub confirmation_action_key: String,
    pub steps: Vec<StepBlueprint>,
    pub expected_window: Option<String>,
}

#[derive(Debug, Default)]
pub struct TemplateRegistry {
    by_intent: HashMap<String, GuidanceTemplate>,
}

impl TemplateRegistry {
    pub fn register(&mut self, template: GuidanceTemplate) {
        self.by_intent.insert(template.intent.clone(), template);
    }

    pub fn get(&self, intent: &str) -> Option<&GuidanceTemplate> {
        self.by_intent.get(intent)
    }

    pub fn known_intents(&self) -> Vec<String> {
        let mut out: Vec<String> = self.by_intent.keys().cloned().collect();
        out.sort();
        out
    }

    pub fn merge_json_dir(&mut self, root: &Path) {
        if !root.exists() {
            return;
        }
        let entries = match std::fs::read_dir(root) {
            Ok(e) => e,
            Err(_) => return,
        };
        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().and_then(|s| s.to_str()) != Some("json") {
                continue;
            }
            let raw = match std::fs::read_to_string(&path) {
                Ok(s) => s,
                Err(_) => continue,
            };
            let parsed: HashMap<String, JsonTemplate> = match serde_json::from_str(&raw) {
                Ok(p) => p,
                Err(err) => {
                    tracing::warn!(target: "roota.templates", "skip {:?}: {err}", path);
                    continue;
                }
            };
            for (intent_name, body) in parsed {
                let steps = body
                    .steps
                    .into_iter()
                    .map(|s| StepBlueprint {
                        action: parse_verb(&s.action),
                        target_query: s.target_query.unwrap_or_else(|| "{target}".into()),
                        instruction_key: s
                            .instruction_key
                            .unwrap_or_else(|| "guidance.click_target".into()),
                        fallback_window: s.fallback_window,
                    })
                    .collect();
                self.register(GuidanceTemplate {
                    intent: intent_name.clone(),
                    confirmation_action_key: body
                        .confirmation_action_key
                        .unwrap_or_else(|| format!("confirm.{intent_name}")),
                    steps,
                    expected_window: body.expected_window,
                });
            }
        }
    }
}

#[derive(Deserialize)]
struct JsonTemplate {
    confirmation_action_key: Option<String>,
    expected_window: Option<String>,
    steps: Vec<JsonStep>,
}

#[derive(Deserialize)]
struct JsonStep {
    action: String,
    target_query: Option<String>,
    instruction_key: Option<String>,
    fallback_window: Option<String>,
}

fn parse_verb(value: &str) -> ActionVerb {
    match value {
        "click" => ActionVerb::Click,
        "double_click" => ActionVerb::DoubleClick,
        "right_click" => ActionVerb::RightClick,
        "type" => ActionVerb::Type,
        _ => ActionVerb::Locate,
    }
}

pub fn default_registry() -> TemplateRegistry {
    let mut r = TemplateRegistry::default();
    let s = |action: ActionVerb, target: &str, key: &str| StepBlueprint {
        action,
        target_query: target.into(),
        instruction_key: key.into(),
        fallback_window: None,
    };
    r.register(GuidanceTemplate {
        intent: "open_folder".into(),
        confirmation_action_key: "confirm.open_folder".into(),
        expected_window: Some("Explorer".into()),
        steps: vec![s(
            ActionVerb::DoubleClick,
            "{target}",
            "guidance.double_click_target",
        )],
    });
    r.register(GuidanceTemplate {
        intent: "move_file".into(),
        confirmation_action_key: "confirm.move_file".into(),
        expected_window: Some("Explorer".into()),
        steps: vec![
            s(ActionVerb::Locate, "{target}", "guidance.locate_target"),
            s(
                ActionVerb::RightClick,
                "{target}",
                "guidance.right_click_target",
            ),
            s(ActionVerb::Click, "Cortar", "guidance.click_target"),
        ],
    });
    r.register(GuidanceTemplate {
        intent: "delete_file".into(),
        confirmation_action_key: "confirm.delete_file".into(),
        expected_window: Some("Explorer".into()),
        steps: vec![
            s(ActionVerb::Click, "{target}", "guidance.click_target"),
            s(ActionVerb::Locate, "Suprimir", "guidance.locate_target"),
        ],
    });
    r.register(GuidanceTemplate {
        intent: "open_browser".into(),
        confirmation_action_key: "confirm.open_browser".into(),
        expected_window: Some("Chrome".into()),
        steps: vec![s(ActionVerb::Locate, "Chrome", "guidance.locate_target")],
    });
    r.register(GuidanceTemplate {
        intent: "search_web".into(),
        confirmation_action_key: "confirm.search_web".into(),
        expected_window: Some("Chrome".into()),
        steps: vec![
            s(ActionVerb::Click, "Nueva pestaña", "guidance.click_target"),
            s(ActionVerb::Type, "Buscar", "guidance.type_in_target"),
        ],
    });
    r.register(GuidanceTemplate {
        intent: "open_url".into(),
        confirmation_action_key: "confirm.open_url".into(),
        expected_window: Some("Chrome".into()),
        steps: vec![
            s(ActionVerb::Click, "Nueva pestaña", "guidance.click_target"),
            s(ActionVerb::Type, "Buscar", "guidance.type_in_target"),
        ],
    });
    r.register(GuidanceTemplate {
        intent: "compose_email".into(),
        confirmation_action_key: "confirm.compose_email".into(),
        expected_window: Some("Gmail".into()),
        steps: vec![s(ActionVerb::Click, "Redactar", "guidance.click_target")],
    });
    r.register(GuidanceTemplate {
        intent: "read_inbox".into(),
        confirmation_action_key: "confirm.read_inbox".into(),
        expected_window: Some("Gmail".into()),
        steps: vec![s(
            ActionVerb::Click,
            "Bandeja de entrada",
            "guidance.click_target",
        )],
    });
    r.register(GuidanceTemplate {
        intent: "reply_message".into(),
        confirmation_action_key: "confirm.reply_message".into(),
        expected_window: Some("Gmail".into()),
        steps: vec![s(ActionVerb::Click, "Responder", "guidance.click_target")],
    });
    r.register(GuidanceTemplate {
        intent: "open_word_document".into(),
        confirmation_action_key: "confirm.open_word_document".into(),
        expected_window: Some("Word".into()),
        steps: vec![s(ActionVerb::Locate, "Word", "guidance.locate_target")],
    });
    r.register(GuidanceTemplate {
        intent: "print_document".into(),
        confirmation_action_key: "confirm.print_document".into(),
        expected_window: Some("Word".into()),
        steps: vec![s(ActionVerb::Click, "Imprimir", "guidance.click_target")],
    });
    r
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn defaults_cover_phase3_intents() {
        let r = default_registry();
        let known = r.known_intents();
        for needed in [
            "open_folder",
            "move_file",
            "delete_file",
            "open_browser",
            "search_web",
            "open_url",
            "compose_email",
            "read_inbox",
            "reply_message",
            "open_word_document",
            "print_document",
        ] {
            assert!(
                known.contains(&needed.to_string()),
                "missing template for {needed}"
            );
        }
    }
}
